### PR TITLE
Replace (most) instances of "catch Throwable" with "catch Exception"

### PR DESCRIPTION
**Pull request by arajasek**
## Notice

It is not allowed to submit your PR to the master branch directly, please submit your PR to the master-pre-merge branch.

## Description

This changeset replaces most of the instances where we catch Throwables in our code with catching Exceptions instead. This addresses Issue #645. 

## Type of change

Insert **x** into the following checkboxes to confirm (eg. [x]):
- [ ] Bug fix.
- [ ] New feature.
- [x] Enhancement.
- [ ] Unit test.
- [ ] Breaking change (a fix or feature that causes existing functionality to not work as expected).
- [ ] Requires documentation update.

## Testing

All existing tests pass. I also ran the kernel with these changes for several hours connected to mainnet with the internal miner running, and there didn't seem to be any issues.

## Verification

Insert **x** into the following checkboxes to confirm (eg. [x]):
- [x] I have self-reviewed my own code and conformed to the style guidelines of this project.
- [x] New and existing tests pass locally with my changes.
- [ ] I have added tests for my fix or feature.
- [ ] I have made appropriate changes to the corresponding documentation.
- [x] My code generates no new warnings.
- [x] Any dependent changes have been made.

### CODE REVIEW

File: modAion/src/org/aion/zero/exceptions/HeaderStructureException.java [hunk 1]
```diff
@@ -56,7 +56,7 @@ public HeaderStructureException(String field, int offset) {
         this.offset = offset;
     }
 
-    public HeaderStructureException(String field, int offset, Throwable cause) {
+    public HeaderStructureException(String field, int offset, Exception cause) {
```
Comment (aion-kelvin) [inline modAion/src/org/aion/zero/exceptions/HeaderStructureException.java]: Throwable should not be caught, for the purpose of avoiding catching Errors that generally can't be handled gracefully.  However, I think it is fine for an Exception subclass to take a Throwable as its cause.  For instance, the class java.lang.Exception allows cause to be a Throwable.  I think we should follow that pattern/design.

**Comment by arajasek** [comment]
Updated with Kelvin's suggestion. 

@AionJayT, the quick StackOverflow research [1][2] I did suggested that "throws throwable" is also a bad idea. Are you sure we should be doing that? I'm happy to defer to you, I'm still new to Java, I just want to double-check.

@aion-kelvin and @AlexandraRoatis, do you have thoughts?

[1] https://stackoverflow.com/questions/12564428/is-throws-throwable-good-practice
[2] https://stackoverflow.com/questions/17934912/should-throwable-ever-be-thrown

**Comment by aion-kelvin** [comment]
I agree with Aayush for this one.  I think the intent (from the design of Java language) of declaring what exceptions are thrown by a method is so that when people write code that use that method, they know what exceptions they need to handle.  In that sense, it would be better to be as narrow as possible when declaring what exceptions it throws.

i.e. if I want to write code that needs to call some method like `void doStuff() throws NumberFormatException` -- well I have a good idea of what problem can arise and how to handle it.  Whereas if I had to call a method like `void doStuff() throws Throwable` then I don't really know how my code is supposed to handle a Throwable.  Seems like I'd have to write exception handling code for every possible exception/error case that could arise.

File: modAionImpl/src/org/aion/zero/impl/AionHub.java [hunk 2]
```diff
@@ -248,7 +248,7 @@ private void loadEventMgr(boolean forTest) {
         prop.put(EventMgrModule.MODULENAME, "org.aion.evtmgr.impl.mgr.EventMgrA0");
         try {
             this.eventMgr = EventMgrModule.getSingleton(prop).getEventMgr();
-        } catch (Throwable e) {
+        } catch (Exception e) {
             genLOG.error("Can not load the Event Manager Module", e.getMessage());
```
Comment (AlexandraRoatis) [inline modAionImpl/src/org/aion/zero/impl/AionHub.java]: additional correction needed here: `getMessage()` may sometime be null. I think it's best to put the whole exception when logging.

File: modAionImpl/src/org/aion/zero/impl/blockchain/AionPendingStateImpl.java [hunk 3]
```diff
@@ -199,7 +199,7 @@ private synchronized void processTxBuffer() {
                     }
                     AionImpl.inst().broadcastTransactions(txs);
                 }
-            } catch (Throwable e) {
+            } catch (Exception e) {
                 LOGGER_TX.error("processTxBuffer throw {}", e.toString());
```
Comment (AlexandraRoatis) [inline modAionImpl/src/org/aion/zero/impl/blockchain/AionPendingStateImpl.java]: `e.toString()` can be replaced by `e`

File: modAionImpl/src/org/aion/zero/impl/blockchain/AionPendingStateImpl.java [hunk 4]
```diff
@@ -1217,7 +1217,7 @@ private void recoverCache() {
         for (byte[] b : pendingCacheTxBytes) {
             try {
                 pendingTx.add(new AionTransaction(b));
-            } catch (Throwable e) {
+            } catch (Exception e) {
                 LOGGER_TX.error("loadingPendingCacheTx error {}", e.toString());
```
Comment (AlexandraRoatis) [inline modAionImpl/src/org/aion/zero/impl/blockchain/AionPendingStateImpl.java]: `e.toString()` can be replaced by `e`

File: modAionImpl/src/org/aion/zero/impl/blockchain/AionPendingStateImpl.java [hunk 5]
```diff
@@ -1257,7 +1257,7 @@ private void recoverPool() {
         for (byte[] b : pendingPoolTxBytes) {
             try {
                 pendingTx.add(new AionTransaction(b));
-            } catch (Throwable e) {
+            } catch (Exception e) {
                 LOGGER_TX.error("loadingCachePendingTx error {}", e.toString());
```
Comment (AlexandraRoatis) [inline modAionImpl/src/org/aion/zero/impl/blockchain/AionPendingStateImpl.java]: `e.toString()` can be replaced by `e`

File: modAionImpl/src/org/aion/zero/impl/sync/TaskImportBlocks.java [hunk 6]
```diff
@@ -274,7 +274,7 @@ private PeerState processBatch(PeerState givenState, List<AionBlock> batch, Stri
                         last = b.getNumber() + 1;
                     }
                 }
-            } catch (Throwable e) {
+            } catch (Exception e) {
                 log.error("<import-block throw> {}", e.toString());
```
Comment (AlexandraRoatis) [inline modAionImpl/src/org/aion/zero/impl/sync/TaskImportBlocks.java]: `e.toString()` can be replaced by `e`

File: modAionImpl/src/org/aion/zero/impl/sync/TaskImportBlocks.java [hunk 7]
```diff
@@ -689,7 +689,7 @@ private int importFromStorage(PeerState state, long first, long last) {
                             // stop importing this queue
                             break;
                         }
-                    } catch (Throwable e) {
+                    } catch (Exception e) {
                         log.error("<import-block throw> {}", e.toString());
```
Comment (AlexandraRoatis) [inline modAionImpl/src/org/aion/zero/impl/sync/TaskImportBlocks.java]: `e.toString()` can be replaced by `e`

File: modAionImpl/src/org/aion/zero/impl/sync/handler/BroadcastTxHandler.java [hunk 8]
```diff
@@ -95,7 +95,7 @@ public void run() {
                 List<AionTransaction> txs = new ArrayList<>();
                 try {
                     txQueue.drainTo(txs);
-                } catch (Throwable e) {
+                } catch (Exception e) {
                     log.error("BufferTask throw{}", e.toString());
```
Comment (AlexandraRoatis) [inline modAionImpl/src/org/aion/zero/impl/sync/handler/BroadcastTxHandler.java]: `e.toString()` can be replaced by `e`

File: modAionImpl/src/org/aion/zero/impl/sync/handler/BroadcastTxHandler.java [hunk 9]
```diff
@@ -145,7 +145,7 @@ public final void receive(int _nodeIdHashcode, String _displayId, final byte[] _
                 }
             }
 
-        } catch (Throwable e) {
+        } catch (Exception e) {
             log.error("BroadcastTxHandler throw {}", e.toString());
```
Comment (AlexandraRoatis) [inline modAionImpl/src/org/aion/zero/impl/sync/handler/BroadcastTxHandler.java]: `e.toString()` can be replaced by `e`

File: modApiServer/src/org/aion/api/server/zmq/HdlrZmq.java [hunk 10]
```diff
@@ -80,7 +80,7 @@ void getTxWait() {
                 return;
             }
 
-        } catch (Throwable e) {
+        } catch (Exception e) {
             // TODO Auto-generated catch block
             LOGGER.error("zmq takeTxWait failed! " + e.getMessage());
```
Comment (AlexandraRoatis) [inline modApiServer/src/org/aion/api/server/zmq/HdlrZmq.java]: `getMessage()` may sometime be null, I suggest replacing `" + e.getMessage());` with `", e);`

File: modEvtMgrImpl/src/org/aion/evtmgr/impl/abs/AbstractHandler.java [hunk 11]
```diff
@@ -92,7 +92,7 @@ public AbstractHandler(int value) {
     public synchronized boolean addEvent(IEvent _evt) {
         try {
             return this.events.add(_evt);
-        } catch (Throwable e) {
+        } catch (Exception e) {
             LOG.error("addEvent exception {}", e.toString());
```
Comment (AlexandraRoatis) [inline modEvtMgrImpl/src/org/aion/evtmgr/impl/abs/AbstractHandler.java]: `e.toString()` can be replaced by `e`

File: modEvtMgrImpl/src/org/aion/evtmgr/impl/abs/AbstractHandler.java [hunk 12]
```diff
@@ -101,7 +101,7 @@ public synchronized boolean addEvent(IEvent _evt) {
     public synchronized boolean removeEvent(IEvent _evt) {
         try {
             return this.events.remove(_evt);
-        } catch (Throwable e) {
+        } catch (Exception e) {
             LOG.error("removeEvent exception {}", e.toString());
```
Comment (AlexandraRoatis) [inline modEvtMgrImpl/src/org/aion/evtmgr/impl/abs/AbstractHandler.java]: `e.toString()` can be replaced by `e`

File: modEvtMgrImpl/src/org/aion/evtmgr/impl/abs/AbstractHandler.java [hunk 13]
```diff
@@ -112,7 +112,7 @@ public void stop() throws InterruptedException {
         interrupt.set(true);
         try {
             this.queue.add(new EventDummy());
-        } catch (Throwable e) {
+        } catch (Exception e) {
             LOG.error("stop exception {}", e.toString());
```
Comment (AlexandraRoatis) [inline modEvtMgrImpl/src/org/aion/evtmgr/impl/abs/AbstractHandler.java]: `e.toString()` can be replaced by `e`

File: modEvtMgrImpl/src/org/aion/evtmgr/impl/abs/AbstractHandler.java [hunk 14]
```diff
@@ -171,7 +171,7 @@ public void start() {
     public void onEvent(IEvent _evt) {
         try {
             this.queue.add(_evt);
-        } catch (Throwable e) {
+        } catch (Exception e) {
             LOG.error("onEvent exception! {}", e.toString());
```
Comment (AlexandraRoatis) [inline modEvtMgrImpl/src/org/aion/evtmgr/impl/abs/AbstractHandler.java]: `e.toString()` can be replaced by `e`

File: modMcf/src/org/aion/mcf/tx/AbstractTxTask.java [hunk 15]
```diff
@@ -71,9 +71,9 @@ public AbstractTxTask(List<TX> _tx, P2P _p2pMgr, Msg _msg) {
             }
 
             return tx;
-        } catch (Throwable th) {
+        } catch (Exception e) {
             // Todo : Log
-            System.out.println(th.getMessage());
+            System.out.println(e.getMessage());
```
Comment (AlexandraRoatis) [inline modMcf/src/org/aion/mcf/tx/AbstractTxTask.java]: probably best to do `e.printStackTrace()` here

File: modP2pImpl/src/org/aion/p2p/impl2/TaskWrite.java [hunk 16]
```diff
@@ -75,7 +75,7 @@ public void run() {
 
             // send outbound event to ioLoop for I/O
             this.ioLoop.write(buf, this.sc);
-        } catch (Throwable e) {
+        } catch (Exception e) {
             System.out.println("<p2p-taskWrite-throw>" + e.toString());
```
Comment (AlexandraRoatis) [inline modP2pImpl/src/org/aion/p2p/impl2/TaskWrite.java]: `e.toString()` can be replaced by `e`

**Pull request merged**